### PR TITLE
Oneline whiteblock parsing fixes

### PR DESCRIPTION
**Pull request by wcjohnson**
Fixes https://github.com/lightscript/lightscript/issues/32

- Allow parsing of single statements in oneline whiteblocks
- Change if and arrow parsing to call new subroutine
- Allow ASI after existing semicolons and in other edge cases
- Add class to startsExpr so `ClassExpression`s don’t get parsed as stmts
- Added fixtures

### CODE REVIEW

**Comment by wcjohnson** [comment]
Hm, what should be done about this:
```js
f() ->
  a = if x: y else: return z
```
```js
function f() {
  const a = x ? y : function () {
    return z;
  }();
  return a;
}
```

This is not technically a parsing problem but the output is semantically incorrect. There is a similar issue in coffeescript which they resolve by raising a syntax error when you put a return inside an expression context.

Should we disallow returns in here at parse time, or defer to compile time?

**Comment by rattrayalex** [comment]
Great question. If we can do at parse time that's better, compile time is ok too. 

My guess is that we could do this by setting `this.state.inFunction` (or w/e) to `false` within such contexts... but there are a number of multiline-expression contexts in LightScript, and unifying them might make sense.

File: src/parser/util.js [hunk 1]
```diff
@@ -65,7 +65,18 @@ pp.isLineBreak = function () {
...
+    (this.hasPlugin("lightscript") && (
+      // LSC oneline statement ASI cases
+      // Allow if x: throw y else: throw z
+      this.match(tt._else) ||
+      this.match(tt._elif) ||
+      // Allow (-> throw new Error)()
+      this.match(tt.parenR) ||
+      // Technically it is legal to insert a ; after a ;.
+      // Allows -> throw new Error; f()
+      this.state.tokens[this.state.tokens.length - 1].type === tt.semi
```
Comment (rattrayalex) [inline src/parser/util.js:78]: hmm... this hints at something being wrong elsewhere...

File: src/plugins/lightscript.js [hunk 2]
```diff
@@ -143,7 +202,7 @@ pp.parseWhiteBlock = function (allowDirectives?, isExpression?) {
   // must start with colon or arrow
   if (isExpression) {
     this.expect(tt.colon);
-    if (!this.isLineBreak()) return this.parseMaybeAssign();
+    if (!this.isLineBreak()) return this.parseWhiteBlockBody(node, indentLevel, true, false);
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: I think a bunch of the above stuff could be replaced with: 

```
if (!this.isLineBreak()) {
  const stmt = this.parseStatement(false);
  return stmt.type === "ExpressionStatement" ? stmt.expression : stmt;
}
```

Comment (rattrayalex) [reply src/plugins/lightscript.js]: Right? Would this obviate the need for `couldBeginStatement` and `parseWhiteBlockBody`?

**Review by rattrayalex — COMMENTED** [review]
haven't had the chance to look through tests yet...

File: src/plugins/lightscript.js (hunk 2 above)
Comment (wcjohnson) [reply src/plugins/lightscript.js]: I will try that, I don't like `couldBeginStatement`. Not sure if there are any additional edge cases there but I can't think of any off hand.

Comment (wcjohnson) [reply src/plugins/lightscript.js]: One possibility is that it will allow things like `break` and `continue` where they don't make any sense, but maybe that is not a parser issue?

Comment (rattrayalex) [reply src/plugins/lightscript.js]: Yeah, I think that's best handled elsewhere since they can happen within a block anyway:

```
x = if true:
  1 + 1
  break
  return
else:
  foo()
  continue
```
etc

though actually I think break and continue might be fine in there...

File: src/parser/util.js (hunk 1 above)
Comment (wcjohnson) [reply src/parser/util.js:78]: The `;` after `;` thing is the result of going down a pretty deep rabbit hole. In JS, `(x) => y; z` is legal (compiles to a `Body` of two `ExpressionStatement`s) and I wanted to maintain consistency.

So that would mean `(x) => throw new Error; z` should compile to a `Body` of two `ExpressionStatements` as well. (In Coffeescript, that actually compiles to one function with two statements in the body)

Ultimately the thing that was preventing that was the absence of this line of code here, so I added it, and it had the result I wanted without breaking JS, so I went with it. Not sure about better options.

Comment (wcjohnson) [inline src/parser/util.js]: The ; after ; thing is the result of going down a pretty deep rabbit hole. In JS, (x) => y; z is legal (compiles to a Body of two ExpressionStatements) and I wanted to maintain consistency.

So that would mean (x) => throw new Error; z should compile to a Body of two ExpressionStatements as well. (In Coffeescript, that actually compiles to one function with two statements in the body)

Ultimately the thing that was preventing that was the absence of this line of code here, so I added it, and it had the result I wanted without breaking JS, so I went with it. Not sure about better options.

File: src/plugins/lightscript.js (hunk 2 above)
Comment (wcjohnson) [reply src/plugins/lightscript.js]: If I understand correctly, the compiler is going to turn that into a ternary, and the break will get wrapped into an iife by `replaceExpressionWithStatements` -- and a break inside an iife by itself I think might be illegal JS. Will test and see what happens.

Comment (wcjohnson) [inline src/plugins/lightscript.js]: Well I got an error message I've never seen before~
```js
x = if true:
  break
```
```
unknown: Unsyntactic break (2:2)
  1 | x = if true:
> 2 |   break
    |   ^
  3 | 
```

Comment (wcjohnson) [reply src/plugins/lightscript.js]:
```js
while true:
  x = if true:
    break
```
```js
while (true) {
  const x = true ? function () {
    break;
  }() : null;
}
```

Comment (wcjohnson) [reply src/plugins/lightscript.js]: Chrome says: ```Uncaught SyntaxError: Illegal break statement``` to that, so it is illegal js

**Comment by wcjohnson** [comment]
Trying to use `parseStatement` as suggested breaks a number of JS fixtures. Most are in the following vein:
```
  index › es2017/async functions/10
  Error: /Users/wcj/dev/lightscript-devkit/packages/babylon-lightscript/test/fixtures/es2017/async-functions/10/actual.js: 26 !== 27 (end/expression/1/body/program)
    runTest (test/utils/runFixtureTests.js:169:13)
    Test.fn (test/utils/runFixtureTests.js:107:20)
```

The issue is that `parseExpressionStatement` is scooping up the `;` into the expression node somehow (which is probably incorrect)

There are a select few stranger ones that seem like legitimate breakages, though:
```js
f(a, async(x, y) => await [x, y], b)
```
```
  index › es2017/async functions/26
  Error: /Users/wcj/dev/lightscript-devkit/packages/babylon-lightscript/test/fixtures/es2017/async-functions/26/actual.js: array length mismatch 3 != 2 (arguments/expression/0/body/program)
    runTest (test/utils/runFixtureTests.js:169:13)
    Test.fn (test/utils/runFixtureTests.js:107:20)
```
The issue here is that lambdas in JS are parsed with `parseMaybeAssign` -- switching to `parseStatement` here is calling `parseExpression` rather than `parseMaybeAssign` -- `parseExpression` can parse sequence expressions -- so finally it interprets that inner function as a lambda returning a `SequenceExpression` as opposed to the outer function call with 3 arguments, which it is in JS.

This seems like quite a can of worms, and my original approach doesn't break any of these things. I'm inclined to stick with it unless you can see another way through.

**Comment by wcjohnson** [comment]
OK, I think this pr is in much better shape now and the case of illegal return etc. inside expressions should be handled better. Should be good to review more closely.

File: src/parser/statement.js [hunk 3]
```diff
@@ -66,7 +66,12 @@ pp.parseStatement = function (declaration, topLevel) {
   // complexity.
 
   switch (starttype) {
-    case tt._break: case tt._continue: return this.parseBreakContinueStatement(node, starttype.keyword);
+    case tt._break:
+      if (this.hasPlugin("lightscript") && isExprContext) this.unexpected(null, "break is illegal inside LightScript expressions");
```
Comment (rattrayalex) [inline src/parser/statement.js]: `Unsyntactic break` is the error babel throws when you try to do what this compiles to, so if anything let's keep that

File: src/parser/statement.js [hunk 4]
```diff
@@ -53,7 +53,7 @@ pp.stmtToDirective = function (stmt) {
 // `if (foo) /blah/.exec(foo)`, where looking at the previous token
 // does not help.
 
-pp.parseStatement = function (declaration, topLevel) {
+pp.parseStatement = function (declaration, topLevel, isExprContext) {
```
Comment (rattrayalex) [inline src/parser/statement.js]: I really don't like adding parameters to preexisting babylon functions...

File: src/plugins/lightscript.js [hunk 5]
```diff
@@ -134,47 +134,80 @@ pp.parseObjectComprehension = function(node) {
...
   }
 
+  this.addExtra(node, "curly", false);
   return this.finishNode(node, "BlockStatement");
 };
 
+pp.parseInlineWhiteBlock = function(node, isExprContext, allowEmptyBody) {
+  if (this.state.type.startsExpr) return this.parseMaybeAssign();
+  // oneline statement case
+  node.body = this.couldBeginStatement() ? [this.parseStatement(true, false, isExprContext)] : [];
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: why is this here? why not just `this.parseStatement`?

File: src/tokenizer/types.js [hunk 6]
```diff
@@ -152,7 +152,7 @@ export const keywords = {
   "new": new KeywordTokenType("new", { beforeExpr, startsExpr }),
   "this": new KeywordTokenType("this", { startsExpr }),
   "super": new KeywordTokenType("super", { startsExpr }),
-  "class": new KeywordTokenType("class"),
+  "class": new KeywordTokenType("class", { startsExpr }),
```
Comment (rattrayalex) [inline src/tokenizer/types.js:155]: This should be in a separate PR (seems like something that should go upstream? would first want to understand why it wasn't here before...)

File: src/plugins/lightscript.js [hunk 7]
```diff
@@ -306,7 +339,23 @@ pp.parseArrowFunctionBody = function (node) {
   this.state.labels = [];
   this.state.inFunction = true;
 
-  node.body = this.parseWhiteBlock(true);
+  const indentLevel = this.state.indentLevel;
+  node.body = this.startNode();
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: unnecessary

Comment (rattrayalex) [reply src/plugins/lightscript.js]: I'd much prefer to keep a single call here as before...

File: src/parser/util.js (hunk 1 above)
Comment (rattrayalex) [reply src/parser/util.js:78]: This really looks wrong to me. None of these things should be here; they belong elsewhere at a call site or something.

File: src/parser/statement.js [hunk 8]
```diff
@@ -79,7 +84,11 @@ pp.parseStatement = function (declaration, topLevel) {
       return this.parseClass(node, true);
 
     case tt._if: return this.parseIfStatement(node);
-    case tt._return: return this.parseReturnStatement(node);
+
+    case tt._return:
+      if (this.hasPlugin("lightscript") && isExprContext) this.unexpected(null, "return is illegal inside LightScript expressions");
```
Comment (rattrayalex) [inline src/parser/statement.js]: Let's instead use a `this.state.inExpressionBlock` or something, and inside `this.parseReturnStatement` / `this.parseBreakContinueStatement` do the check. 

Then we won't need to pass around `isExprContext` like this or modify `this.parseStatement` itself at all.

File: src/plugins/lightscript.js (hunk 5 above)
Comment (rattrayalex) [reply src/plugins/lightscript.js]: Have you tried the approach I outlined in gitter? Just `parseStatement` and extract the expression (if relevant) and/or raise on illegal things thereafter? 

I haven't fiddled, so I'm not sure what problems it might cause, but my hunch is it would reduce the LOC diff of this PR by a factor of 3-5...

File: src/plugins/lightscript.js [hunk 9]
```diff
@@ -134,47 +134,80 @@ pp.parseObjectComprehension = function(node) {
...
+pp.finishWhiteBlock = function(node, allowEmptyBody) {
   if (!allowEmptyBody && !node.body.length) {
     this.unexpected(node.start, "Expected an Indent or Statement");
   }
 
+  this.addExtra(node, "curly", false);
   return this.finishNode(node, "BlockStatement");
 };
 
+pp.parseInlineWhiteBlock = function(node, isExprContext, allowEmptyBody) {
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: mixed feelings about breaking these out in general, but I see the appeal... should probably be called Oneline not Inline though...

File: src/plugins/lightscript.js (hunk 5 above)
Comment (wcjohnson) [reply src/plugins/lightscript.js]: Yes, I did try that approach as you suggested. I left an extensive comment about it which I assume got eaten or something?

It causes several dozen JS fixtures to break. The most interesting breakages involve ambiguities between `SequenceExpression`s and function argument lists. The breakages are caused by `parseStatement` going down a deeper parsing pathway than `parseMaybeAssign`, along which comma-separated expressions get turned into `SequenceExpression`s instead of unwinding the recursive descent.

**Comment by wcjohnson** [comment]
Sounds like you may have some better ideas than I do about how to do this, so I think I'll kick it back your way. Branch will be open if you want any of the code.

**Comment by rattrayalex** [comment]
Hey, sorry I didn't see that earlier comment. Thanks for taking the time to explain

File: src/parser/util.js (hunk 1 above)
Comment (rattrayalex) [reply src/parser/util.js:78]: I think this should be able to go away once you rebase on master, does that sound right? (guessing you just haven't gotten to that yet)

Comment (wcjohnson) [reply src/parser/util.js:78]: It's not clear to me how that would happen. How is `(-> throw new Error)()` going to compile? It calls `pp.semicolon()` after parsing the throw, but there isn't a semicolon, EOL, or brace there...

Unless I misunderstand the ASI portion of the parser in some fundamental way, I believe this change is necessary.

**Comment by wcjohnson** [comment]
@rattrayalex 

I applied the fixes for empty whiteblocks to this branch. Because of the divergences, the changes were basically not directly mergeable.

I think you should consider reverting the merge from last night and merging this branch instead (or just resolve conflicts with my side)

File: src/parser/util.js (hunk 1 above)
Comment (rattrayalex) [reply src/parser/util.js:78]: ah, gotcha. Wasn't clear to me earlier that that was the problem. 

Do you know off the top of your head if it's just `throw` or if this problem exists across many desirably legal statements?

One possibility would be to modify `parseThrowStatement` to be "softer" than just `this.semicolon()`. 

Another would be to parse `throw` as an expression generally, which I think could make sense as a looser lightscript semantic (would just need to ensure it's sometimes wrapped in an iife in compiler). 

The last thing on my mind here is that the cases you've added here do indeed make sense in this context and quite probably others – and I'm thinking about whether they should be enabled conditionally and at what layer of abstraction. 

Just want to be thoughtful about this. I'm hopeful to be able to play around with it myself later today, and there's a good chance I'll end up in solid agreement that the current implementation is the best path forward.

**Comment by rattrayalex** [comment]
Cool, I can try to resolve the conflicts

File: src/plugins/lightscript.js [hunk 10]
```diff
@@ -134,40 +134,45 @@ pp.parseObjectComprehension = function(node) {
   return this.finishNode(node, "ObjectComprehension");
 };
 
-// c/p parseBlock
+pp.finishWhiteBlock = function(node) {
+  if (!node.body.length) {
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: I think this check can now be moved to `parseMultilineWhiteBlock`

File: src/plugins/lightscript.js [hunk 11]
```diff
@@ -134,40 +134,45 @@ pp.parseObjectComprehension = function(node) {
...
+
+  this.addExtra(node, "curly", false);
+  return this.finishNode(node, "BlockStatement");
+};
 
-pp.parseWhiteBlock = function (allowDirectives?, isExpression?) {
+pp.parseInlineWhiteBlock = function(node) {
+  if (this.state.type.startsExpr) return this.parseMaybeAssign();
+  // oneline statement case
+  node.body = [this.parseStatement(true)];
```
Comment (rattrayalex) [inline src/plugins/lightscript.js]: shouldn't be in a list

File: src/plugins/lightscript.js (hunk 10 above)
Comment (rattrayalex) [reply src/plugins/lightscript.js]: ... in which case I'm not sure this method is necessary (up to you)

File: src/plugins/lightscript.js (hunk 7 above)
Comment (rattrayalex) [reply src/plugins/lightscript.js]: nevermind about the single call, but `node.body = this.startNode();` here can definitely be removed

File: src/tokenizer/types.js (hunk 6 above)
Comment (rattrayalex) [reply src/tokenizer/types.js:155]: eh, this is fine in here. I think it makes sense

**Comment by rattrayalex** [comment]
@wcjohnson would you be willing to squash commits?

**Comment by wcjohnson** [comment]
@rattrayalex Yes, when I get a moment I will clean up the comments you left and then I can squash after a final go ahead

File: src/plugins/lightscript.js (hunk 11 above)
Comment (wcjohnson) [reply src/plugins/lightscript.js]: Are you sure about this? The node gets turned into a `BlockStatement` by `finishWhiteBlock`, and babel-types says those have to have bodies that are arrays:

```
defineType("BlockStatement", {
  builder: ["body", "directives"],
  visitor: ["directives", "body"],
  fields: {
    directives: {
      validate: chain(assertValueType("array"), assertEach(assertNodeType("Directive"))),
      default: [],
    },
    body: {
      validate: chain(assertValueType("array"), assertEach(assertNodeType("Statement"))),
    },
  },
  aliases: ["Scopable", "BlockParent", "Block", "Statement"],
});
```

Comment (wcjohnson) [reply src/plugins/lightscript.js]: or I guess you're saying just return the parsed statement and don't wrap it at all?

Comment (wcjohnson) [reply src/plugins/lightscript.js]: I tried this change and there is a problem. `parseArrowFunctionBody` is doing a fixup here: https://github.com/wcjohnson/babylon-lightscript/blob/upstream/oneline-whiteblocks/src/plugins/lightscript.js#L328

Because the parser is no longer producing `BlockStatement`s after this change, functions are getting false-flagged as `expression=true` when they aren't expression bodies.

**Comment by wcjohnson** [comment]
@rattrayalex 

Upon testing, removing that `node.body = this.startNode()` broke things and was incorrect. We need to start a node there because the whiteblock body parser uses that node as the place to insert the body, and that location includes the arrow token.

I clarified the code a little bit but semantically it's the same.

**Comment by wcjohnson** [comment]
@rattrayalex Comments have been addressed, let me know if anything else needs doing before the squash

**Comment by wcjohnson** [comment]
@rattrayalex 
I think one of my comments got eaten. Re your comments on:
```js
  // oneline statement case
  node.body = [this.parseStatement(true)];
  node.directives = [];
  this.addExtra(node, "curly", false);
  return this.finishNode(node, "BlockStatement");
```

You suggested dropping the array there. I'm not sure that's correct. According to `babel-types`, a `BlockStatement` requires an array as its body. We could cut out the `BlockStatement` and just return `parseStatement` but this breaks two things:

1) There is a fixup later down the line in `parseArrowBody` that turns anything not a `BlockStatement` as an expression. Returning a vanilla statement here breaks that check and breaks a lot of function body fixtures by falsely flagging the bodies as expressions.

2) Calling `parseStatement` there causes the arrow token not to be included in the arrow body's source location. I am not sure whether that's right or not, but heretofore, we have been parsing the arrow token as being part of the arrow body when it's a whiteblock.

**Comment by rattrayalex** [comment]
Cool, thanks for explaining! That definitely makes sense. I'll give another look but anticipate being 👍

File: src/plugins/lightscript.js [hunk 12]
```diff
@@ -301,7 +302,23 @@ pp.parseArrowFunctionBody = function (node) {
   this.state.labels = [];
   this.state.inFunction = true;
 
-  node.body = this.parseWhiteBlock(true);
+  const indentLevel = this.state.indentLevel;
+  const nodeAtArrow = this.startNode();
```
Comment (rattrayalex) [inline src/plugins/lightscript.js:306]: ah, cool, thanks – that is clearer

File: test/fixtures/lightscript/arrow-methods/obj-skinny-oneline-stmt/actual.js [hunk 13]
```diff
@@ -0,0 +1,3 @@
+x = {
+  method() -> for idx i in Array(10): i
```
Comment (rattrayalex) [inline test/fixtures/lightscript/arrow-methods/obj-skinny-oneline-stmt/actual.js:2]: jw was there a special case that methods hit?

**Review by rattrayalex — APPROVED** [review]
This looks good, ready for squash! Thanks!

I do hope to cleanup the `canInsertSemicolon` bit in the future...

**Comment by wcjohnson** [comment]
Squashing this is actually a little tricky because it straddles the merge and unmerge of babylon 7. Let me see what I can do.

**Comment by wcjohnson** [comment]
@rattrayalex OK, that was an ugly rebase but I think I got it right, should be squashed now.

File: test/fixtures/lightscript/arrow-methods/obj-skinny-oneline-stmt/actual.js (hunk 13 above)
Comment (wcjohnson) [reply test/fixtures/lightscript/arrow-methods/obj-skinny-oneline-stmt/actual.js:2]: I don't think so; this is probably a redundant fixture.

**Comment by rattrayalex** [comment]
Woohoo! This was a heavy lift. Great work @wcjohnson and thanks for putting up with all my (sometimes ignorant) asks.
